### ISSUE TITLE
refactor(web): context operations should use a sliding window

### ISSUE THREAD

**Issue by jahorton** [keymanapp/keyman#5258]
> ...
> 
> Still, it does just move the goalposts. As expected, when I tested with longer documents, I observed linear growth in script time. So this still fails with long documents -- and not all that long really, with just over 1 page of text in a note, I saw 160msec/key event which is barely tolerable. This is on a relatively fast device (for Android), so slower devices would probably be completely unusable with that much content.
> 
> The big problem in this whole algorithm is that it is ever scanning the whole document. That's not limited to this one function, but to the whole approach, I think. That seems completely unnecessary: we're never going to have a keyboard that affects more than the last few characters.
> 
> We should be working with a sliding window, say 128 code _units_ either side of the caret. Is this possible to achieve? We may still need to iterate through the document once to get our relative code _point_ base for the sliding window but that should be done just once for a keystroke and would then be relatively inconsequential.
> 
> I know that all our indexing is by document position, so it may be a significant challenge to rewrite it. But improved algorithms are only going to take us so far and I think we may need to consider this for a full fix.

As noted, this is important... but is also _definitely_ not trivial.

In terms of context passed through to the `KeyboardProcessor`, there would be little problem.  As noted, keystrokes don't ever have that wide a range of effect.

The major pain point would be retooling how suggestions and reversions are applied.  The process for both involves comparing the "before" snapshot with the current context state, which gets tricky when context is represented by sliding windows.  It does seem _possible_, though.

I do think we should have a "design" round before making such a change.  There are a number of ramifications to think through, especially given how critical context synchronization is when embedded within the mobile apps.

_Originally posted by @jahorton in https://github.com/keymanapp/keyman/issues/5248#issuecomment-858198679_

**Comment by mcdurdin** [comment]
Given we are not observing current performance issues here, let's wait on this.

**Comment by ermshiperete** [comment]
Related to #5312 ?

**Comment by jahorton** [comment]
> Related to #5312 ?

Nah, that has nothing to do with predictive text and everything to do with how the `TouchAliasElement` operates.  It's not exactly an optimized word-processing element.

**Comment by mcdurdin** [comment]
Although ... good to think about both areas in tandem!

**Comment by jahorton** [comment]
This is the sort of change that should happen early in a development cycle, by the way.  Lots of chances for things to go wrong if something is broken, and I'm not convinced that all possible side-effects would be obvious.

**Comment by jahorton** [comment]
Before taking in extra constraints specific to our use cases, I believe a major part of this problem can be rephrased as the ["longest common substring" problem](https://en.wikipedia.org/wiki/Longest_common_substring).  When the context window slides, we expect most of the context string to remain intact.

An extra constraint we can impose:
- We expect the "longest common substring" between (old context + recent keystroke transform) and (post-slide context) to lie on one of the two borders for _both_ context variants.
    - We _do_ need to factor in the post-keystroke transform in order to guarantee this.
    - If new text is added:  the "new context" window would "slide right"
       - right-hand side of "old context" should match left-hand side of "new context"
          - left-hand of "old context": holds chars that were previously out-of-window
          - right-hand of "new context": holds newly-typed chars
    - If old text is deleted, pre-caret:  the "new context" window would "slide left"
      - left-hand side of "old context" should match the right-hand side of "new context"
          - right-hand of "old context": holds deleted chars
          - left-hand of "new context": holds chars that were previously out-of-window

Note that this one constraint should _dramatically_ reduce the runtime complexity of a would-be dynamic programming approach - we only need to check for matching substrings on the border of both ends.  This should give us a O(`n`+`m`) time complexity, with `n` and `m` being the length of the two strings, rather than the O(`n`*`m`) we'd get if the longest substrings could start at any index within both contexts.

**Comment by jahorton** [comment]
#10662 (and its state as of commit https://github.com/keymanapp/keyman/pull/10662/commits/b1adb857b3facb8b37a7c1c1cd73107da1813556) may be a useful reference if and when we decide to tackle this issue.  It should be particularly be relevant to the most recent comment above this one.

That said, I imagine that the outcome of the discussion on https://github.com/keymanapp/keyman/pull/10662#discussion_r1480975734 will be relevant to whatever approach ought be taken for sliding window operations as well.

**Comment by jahorton** [comment]
The current work in #14476 and some of the PRs leading up to it also look pretty relevant for this task - the worker's sliding context-window tracking functionality is receiving some upgrades, which should be pretty related to what would be needed in the main Web engine.
